mist-core/src/main/java/edu/snu/mist/core/master/lb/allocation/ApplicationAwareQueryAllocationManager.java

```diff
@@ -101,36 +100,38 @@ private String getRandomTask(final List<String> allTaskList) {
   @Override
   public IPAddress getAllocatedTask(final String appId) {
-    if (!appTaskListMap.containsKey(appId)) {
-      appTaskListMap.putIfAbsent(appId, new ArrayList<>());
-    }
-
-    final List<String> taskList = appTaskListMap.get(appId);
-    synchronized (taskList) {
-      if (taskList.isEmpty()) {
-        List<String> allTaskList;
-        do {
-          allTaskList = taskStatsMap.getTaskList();
-        } while (allTaskList.isEmpty());
-        final String selectedTask = getRandomTask(allTaskList);
-        taskList.add(selectedTask);
-        return taskAddressInfoMap.getClientToTaskAddress(selectedTask);
-      } else {
-        final String selectedTask = getRandomTask(taskList);
-        final double selectedTaskLoad = taskStatsMap.get(selectedTask).getTaskLoad();
-        if (selectedTaskLoad > overloadedTaskThreshold) {
-          // All the tasks are overloaded. Allocate to a new task.
-          final List<String> remainingList = taskStatsMap.getTaskList();
-          remainingList.removeAll(taskList);
-          if (!remainingList.isEmpty()) {
-            final String taskCandidate = getRandomTask(remainingList);
-            if (taskStatsMap.get(taskCandidate).getTaskLoad() < overloadedTaskThreshold) {
-              taskList.add(taskCandidate);
-              return taskAddressInfoMap.getClientToTaskAddress(taskCandidate);
+    // Iterate until success.
+    while (true) {
+      try {
+        final List<String> taskList = appTaskListMap.getTaskListForApp(appId);
+          if (taskList == null) {
+            List<String> allTaskList;
+            do {
+              allTaskList = taskStatsMap.getTaskList();
+            } while (allTaskList.isEmpty());
+            final String selectedTask = getRandomTask(allTaskList);
+            appTaskListMap.addTaskToApp(appId, selectedTask);
+            return taskAddressInfoMap.getClientToTaskAddress(selectedTask);
+          } else {
+            final String selectedTask = getRandomTask(taskList);
+            final double selectedTaskLoad = taskStatsMap.get(selectedTask).getTaskLoad();
+            if (selectedTaskLoad > overloadedTaskThreshold) {
+              // All the tasks are overloaded. Allocate to a new task.
+              final List<String> remainingList = taskStatsMap.getTaskList();
+              remainingList.removeAll(taskList);
+              if (!remainingList.isEmpty()) {
+                final String taskCandidate = getRandomTask(remainingList);
+                if (taskStatsMap.get(taskCandidate).getTaskLoad() < overloadedTaskThreshold) {
+                  appTaskListMap.addTaskToApp(appId, taskCandidate);
+                  return taskAddressInfoMap.getClientToTaskAddress(taskCandidate);
+                }
+              }
             }
+            return taskAddressInfoMap.getClientToTaskAddress(selectedTask);
           }
-        }
-        return taskAddressInfoMap.getClientToTaskAddress(selectedTask);
+      } catch (final Exception e) {
```

Comment:
Where does this exception can occur? it would be good to reduce the area of `try{ ...}` clause.